torch/testing/_internal/common_methods_invocations.py

```diff
@@ -7067,6 +7067,41 @@ def _gather(shape, index_dim, max_indices):
             )
         )
 
+    # Sample inputs to test edge cases for backward
+    # Check that gradients are propagated correctly for prod when zeros in self/src are reduced
+    if requires_grad and reduce == 'prod':
+        # This sample tests gradients for the following cases
+        # (a) 1 zero reduced (from src (self[0, 1]), from self (self[0, 0]))
+        # (b) 2 zeros reduced (1 from src and 1 from self (self[1, 0]), 2 from src (self[1, 1]))
+        # (c) > 2 zeros reduced (2 from src, 1 from self (self[2, 0]))
+        # (d) no zeros reduced (self([2, 1]))
+        input = torch.tensor([[0, 13], [0, 17], [0, 19]], dtype=dtype, device=device, requires_grad=requires_grad)
+        src = torch.tensor([[0, 1, 2, 3], [0, 4, 0, 0], [0, 0, 5, 6]], dtype=dtype, device=device, requires_grad=requires_grad)
+        idx = torch.tensor([[1, 1, 0, 0], [0, 0, 1, 1], [0, 0, 0, 1]], dtype=torch.long, device=device)
+
+        sample_inputs.append(
```

Comment:
fyi the sample inputs functions can be actual generators now and you can just yield SampleInputs as you create them instead of collecting them and return them in an iterable